Improve traceback/error message for assertion utils

### Description

Here's a sample test suite, as context for the discussion below:

```python
# test_asserts.py
import polars as pl
from polars.testing import assert_frame_equal, assert_series_equal

def test_series() -> None:
    s1 = pl.Series([1, 2, 3])
    s2 = pl.Series([1, 2, 4])
    assert_series_equal(s1, s2)

def test_df() -> None:
    df1 = pl.DataFrame({"a": [1, 2, 3]})
    df2 = pl.DataFrame({"a": [1, 2, 4]})
    assert_frame_equal(df1, df2)
```

### Tracebacks

Tracebacks are currently very unwieldy. Here's the output of running `pytest` on the file above:

<details>

```
======================================================================== test session starts =========================================================================
platform linux -- Python 3.11.2, pytest-7.4.0, pluggy-1.3.0
rootdir: /home/stijn/code/polars/py-polars
configfile: pyproject.toml
plugins: cov-4.1.0, hypothesis-6.87.1, xdist-3.3.1
collected 2 items                                                                                                                                                    

test_asserts.py FF                                                                                                                                             [100%]

============================================================================== FAILURES ==============================================================================
____________________________________________________________________________ test_series _____________________________________________________________________________

    def test_series() -> None:
        s1 = pl.Series([1, 2, 3])
        s2 = pl.Series([1, 2, 4])
>       assert_series_equal(s1, s2)

test_asserts.py:8: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
polars/testing/asserts.py:268: in assert_series_equal
    _assert_series_inner(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = shape: (1,)
Series: '' [i64]
[
        3
], right = shape: (1,)
Series: '' [i64]
[
        4
]

    def _assert_series_inner(
        left: Series,
        right: Series,
        *,
        check_dtype: bool,
        check_exact: bool,
        atol: float,
        rtol: float,
        nans_compare_equal: bool,
        categorical_as_str: bool,
    ) -> None:
        """Compare Series dtype + values."""
        if check_dtype and left.dtype != right.dtype:
            raise_assert_detail("Series", "dtype mismatch", left.dtype, right.dtype)
    
        if left.null_count() != right.null_count():
            raise_assert_detail(
                "Series", "null_count is not equal", left.null_count(), right.null_count()
            )
    
        if categorical_as_str and left.dtype == Categorical:
            left = left.cast(Utf8)
            right = right.cast(Utf8)
    
        # create mask of which (if any) values are unequal
        unequal = left.ne_missing(right)
    
        # handle NaN values (which compare unequal to themselves)
        comparing_float_dtypes = left.dtype in FLOAT_DTYPES and right.dtype in FLOAT_DTYPES
        if unequal.any() and nans_compare_equal:
            # when both dtypes are scalar floats
            if comparing_float_dtypes:
                unequal = unequal & ~(
                    (left.is_nan() & right.is_nan()).fill_null(F.lit(False))
                )
        if comparing_float_dtypes and not nans_compare_equal:
            unequal = unequal | left.is_nan() | right.is_nan()
    
        # check nested dtypes in separate function
        if left.dtype.is_nested or right.dtype.is_nested:
            if _assert_series_nested(
                left=left.filter(unequal),
                right=right.filter(unequal),
                check_dtype=check_dtype,
                check_exact=check_exact,
                atol=atol,
                rtol=rtol,
                nans_compare_equal=nans_compare_equal,
                categorical_as_str=categorical_as_str,
            ):
                return
    
        try:
            can_be_subtracted = hasattr(dtype_to_py_type(left.dtype), "__sub__")
        except NotImplementedError:
            can_be_subtracted = False
    
        check_exact = (
            check_exact or not can_be_subtracted or left.is_boolean() or left.is_temporal()
        )
    
        # assert exact, or with tolerance
        if unequal.any():
            if check_exact:
                raise_assert_detail(
                    "Series",
                    "exact value mismatch",
                    left=list(left),
                    right=list(right),
                )
            else:
                # apply check with tolerance (to the known-unequal matches).
                left, right = left.filter(unequal), right.filter(unequal)
    
                if all(tp in UNSIGNED_INTEGER_DTYPES for tp in (left.dtype, right.dtype)):
                    # avoid potential "subtract-with-overflow" panic on uint math
                    s_diff = Series(
                        "diff", [abs(v1 - v2) for v1, v2 in zip(left, right)], dtype=UInt64
                    )
                else:
                    s_diff = (left - right).abs()
    
                mismatch, nan_info = False, ""
                if ((s_diff > (atol + rtol * right.abs())).sum() != 0) or (
                    left.is_null() != right.is_null()
                ).any():
                    mismatch = True
                elif comparing_float_dtypes:
                    # note: take special care with NaN values.
                    # if NaNs don't compare as equal, any NaN in the left Series is
                    # sufficient for a mismatch because the if condition above already
                    # compares the null values.
                    if not nans_compare_equal and left.is_nan().any():
                        nan_info = " (nans_compare_equal=False)"
                        mismatch = True
                    elif (left.is_nan() != right.is_nan()).any():
                        nan_info = f" (nans_compare_equal={nans_compare_equal})"
                        mismatch = True
    
                if mismatch:
>                   raise_assert_detail(
                        "Series",
                        f"value mismatch{nan_info}",
                        left=list(left),
                        right=list(right),
                    )
E                   AssertionError: Series are different (value mismatch)
E                   [left]:  [3]
E                   [right]: [4]

polars/testing/asserts.py:444: AssertionError
______________________________________________________________________________ test_df _______________________________________________________________________________

left = shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
└─────┘
right = shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 4   │
└─────┘

    def assert_frame_equal(
        left: DataFrame | LazyFrame,
        right: DataFrame | LazyFrame,
        *,
        check_row_order: bool = True,
        check_column_order: bool = True,
        check_dtype: bool = True,
        check_exact: bool = False,
        rtol: float = 1.0e-5,
        atol: float = 1.0e-8,
        nans_compare_equal: bool = True,
        categorical_as_str: bool = False,
    ) -> None:
        """
        Raise detailed AssertionError if `left` does NOT equal `right`.
    
        Parameters
        ----------
        left
            the DataFrame to compare.
        right
            the DataFrame to compare with.
        check_row_order
            if False, frames will compare equal if the required rows are present,
            irrespective of the order in which they appear; as this requires
            sorting, you cannot set on frames that contain unsortable columns.
        check_column_order
            if False, frames will compare equal if the required columns are present,
            irrespective of the order in which they appear.
        check_dtype
            if True, data types need to match exactly.
        check_exact
            if False, test if values are within tolerance of each other
            (see `rtol` & `atol`).
        rtol
            relative tolerance for inexact checking. Fraction of values in `right`.
        atol
            absolute tolerance for inexact checking.
        nans_compare_equal
            if your assert/test requires float NaN != NaN, set this to False.
        categorical_as_str
            Cast categorical columns to string before comparing. Enabling this helps
            compare DataFrames that do not share the same string cache.
    
        Examples
        --------
        >>> from polars.testing import assert_frame_equal
        >>> df1 = pl.DataFrame({"a": [1, 2, 3]})
        >>> df2 = pl.DataFrame({"a": [2, 3, 4]})
        >>> assert_frame_equal(df1, df2)  # doctest: +SKIP
        AssertionError: Values for column 'a' are different.
        """
        collect_input_frames = isinstance(left, LazyFrame) and isinstance(right, LazyFrame)
        if collect_input_frames:
            objs = "LazyFrames"
        elif isinstance(left, DataFrame) and isinstance(right, DataFrame):
            objs = "DataFrames"
        else:
            raise_assert_detail("Inputs", "unexpected input types", type(left), type(right))
    
        if left_not_right := [c for c in left.columns if c not in right.columns]:
            raise AssertionError(
                f"columns {left_not_right!r} in left frame, but not in right"
            )
    
        if right_not_left := [c for c in right.columns if c not in left.columns]:
            raise AssertionError(
                f"columns {right_not_left!r} in right frame, but not in left"
            )
    
        if check_column_order and left.columns != right.columns:
            raise AssertionError(
                f"columns are not in the same order:\n{left.columns!r}\n{right.columns!r}"
            )
    
        if collect_input_frames:
            if check_dtype:  # check this _before_ we collect
                left_schema, right_schema = left.schema, right.schema
                if left_schema != right_schema:
                    raise_assert_detail(
                        objs, "lazy schemas are not equal", left_schema, right_schema
                    )
            left, right = left.collect(), right.collect()  # type: ignore[union-attr]
    
        if left.shape[0] != right.shape[0]:  # type: ignore[union-attr]
            raise_assert_detail(objs, "length mismatch", left.shape, right.shape)  # type: ignore[union-attr]
    
        if not check_row_order:
            try:
                left = left.sort(by=left.columns)
                right = right.sort(by=left.columns)
            except ComputeError as exc:
                raise InvalidAssert(
                    "cannot set `check_row_order=False` on frame with unsortable columns"
                ) from exc
    
        # note: does not assume a particular column order
        for c in left.columns:
            try:
>               _assert_series_inner(
                    left[c],  # type: ignore[arg-type, index]
                    right[c],  # type: ignore[arg-type, index]
                    check_dtype=check_dtype,
                    check_exact=check_exact,
                    atol=atol,
                    rtol=rtol,
                    nans_compare_equal=nans_compare_equal,
                    categorical_as_str=categorical_as_str,
                )

polars/testing/asserts.py:124: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = shape: (1,)
Series: 'a' [i64]
[
        3
], right = shape: (1,)
Series: 'a' [i64]
[
        4
]

    def _assert_series_inner(
        left: Series,
        right: Series,
        *,
        check_dtype: bool,
        check_exact: bool,
        atol: float,
        rtol: float,
        nans_compare_equal: bool,
        categorical_as_str: bool,
    ) -> None:
        """Compare Series dtype + values."""
        if check_dtype and left.dtype != right.dtype:
            raise_assert_detail("Series", "dtype mismatch", left.dtype, right.dtype)
    
        if left.null_count() != right.null_count():
            raise_assert_detail(
                "Series", "null_count is not equal", left.null_count(), right.null_count()
            )
    
        if categorical_as_str and left.dtype == Categorical:
            left = left.cast(Utf8)
            right = right.cast(Utf8)
    
        # create mask of which (if any) values are unequal
        unequal = left.ne_missing(right)
    
        # handle NaN values (which compare unequal to themselves)
        comparing_float_dtypes = left.dtype in FLOAT_DTYPES and right.dtype in FLOAT_DTYPES
        if unequal.any() and nans_compare_equal:
            # when both dtypes are scalar floats
            if comparing_float_dtypes:
                unequal = unequal & ~(
                    (left.is_nan() & right.is_nan()).fill_null(F.lit(False))
                )
        if comparing_float_dtypes and not nans_compare_equal:
            unequal = unequal | left.is_nan() | right.is_nan()
    
        # check nested dtypes in separate function
        if left.dtype.is_nested or right.dtype.is_nested:
            if _assert_series_nested(
                left=left.filter(unequal),
                right=right.filter(unequal),
                check_dtype=check_dtype,
                check_exact=check_exact,
                atol=atol,
                rtol=rtol,
                nans_compare_equal=nans_compare_equal,
                categorical_as_str=categorical_as_str,
            ):
                return
    
        try:
            can_be_subtracted = hasattr(dtype_to_py_type(left.dtype), "__sub__")
        except NotImplementedError:
            can_be_subtracted = False
    
        check_exact = (
            check_exact or not can_be_subtracted or left.is_boolean() or left.is_temporal()
        )
    
        # assert exact, or with tolerance
        if unequal.any():
            if check_exact:
                raise_assert_detail(
                    "Series",
                    "exact value mismatch",
                    left=list(left),
                    right=list(right),
                )
            else:
                # apply check with tolerance (to the known-unequal matches).
                left, right = left.filter(unequal), right.filter(unequal)
    
                if all(tp in UNSIGNED_INTEGER_DTYPES for tp in (left.dtype, right.dtype)):
                    # avoid potential "subtract-with-overflow" panic on uint math
                    s_diff = Series(
                        "diff", [abs(v1 - v2) for v1, v2 in zip(left, right)], dtype=UInt64
                    )
                else:
                    s_diff = (left - right).abs()
    
                mismatch, nan_info = False, ""
                if ((s_diff > (atol + rtol * right.abs())).sum() != 0) or (
                    left.is_null() != right.is_null()
                ).any():
                    mismatch = True
                elif comparing_float_dtypes:
                    # note: take special care with NaN values.
                    # if NaNs don't compare as equal, any NaN in the left Series is
                    # sufficient for a mismatch because the if condition above already
                    # compares the null values.
                    if not nans_compare_equal and left.is_nan().any():
                        nan_info = " (nans_compare_equal=False)"
                        mismatch = True
                    elif (left.is_nan() != right.is_nan()).any():
                        nan_info = f" (nans_compare_equal={nans_compare_equal})"
                        mismatch = True
    
                if mismatch:
>                   raise_assert_detail(
                        "Series",
                        f"value mismatch{nan_info}",
                        left=list(left),
                        right=list(right),
                    )
E                   AssertionError: Series are different (value mismatch)
E                   [left]:  [3]
E                   [right]: [4]

polars/testing/asserts.py:444: AssertionError

The above exception was the direct cause of the following exception:

    def test_df() -> None:
        df1 = pl.DataFrame({"a": [1, 2, 3]})
        df2 = pl.DataFrame({"a": [1, 2, 4]})
>       assert_frame_equal(df1, df2)

test_asserts.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
└─────┘
right = shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 4   │
└─────┘

    def assert_frame_equal(
        left: DataFrame | LazyFrame,
        right: DataFrame | LazyFrame,
        *,
        check_row_order: bool = True,
        check_column_order: bool = True,
        check_dtype: bool = True,
        check_exact: bool = False,
        rtol: float = 1.0e-5,
        atol: float = 1.0e-8,
        nans_compare_equal: bool = True,
        categorical_as_str: bool = False,
    ) -> None:
        """
        Raise detailed AssertionError if `left` does NOT equal `right`.
    
        Parameters
        ----------
        left
            the DataFrame to compare.
        right
            the DataFrame to compare with.
        check_row_order
            if False, frames will compare equal if the required rows are present,
            irrespective of the order in which they appear; as this requires
            sorting, you cannot set on frames that contain unsortable columns.
        check_column_order
            if False, frames will compare equal if the required columns are present,
            irrespective of the order in which they appear.
        check_dtype
            if True, data types need to match exactly.
        check_exact
            if False, test if values are within tolerance of each other
            (see `rtol` & `atol`).
        rtol
            relative tolerance for inexact checking. Fraction of values in `right`.
        atol
            absolute tolerance for inexact checking.
        nans_compare_equal
            if your assert/test requires float NaN != NaN, set this to False.
        categorical_as_str
            Cast categorical columns to string before comparing. Enabling this helps
            compare DataFrames that do not share the same string cache.
    
        Examples
        --------
        >>> from polars.testing import assert_frame_equal
        >>> df1 = pl.DataFrame({"a": [1, 2, 3]})
        >>> df2 = pl.DataFrame({"a": [2, 3, 4]})
        >>> assert_frame_equal(df1, df2)  # doctest: +SKIP
        AssertionError: Values for column 'a' are different.
        """
        collect_input_frames = isinstance(left, LazyFrame) and isinstance(right, LazyFrame)
        if collect_input_frames:
            objs = "LazyFrames"
        elif isinstance(left, DataFrame) and isinstance(right, DataFrame):
            objs = "DataFrames"
        else:
            raise_assert_detail("Inputs", "unexpected input types", type(left), type(right))
    
        if left_not_right := [c for c in left.columns if c not in right.columns]:
            raise AssertionError(
                f"columns {left_not_right!r} in left frame, but not in right"
            )
    
        if right_not_left := [c for c in right.columns if c not in left.columns]:
            raise AssertionError(
                f"columns {right_not_left!r} in right frame, but not in left"
            )
    
        if check_column_order and left.columns != right.columns:
            raise AssertionError(
                f"columns are not in the same order:\n{left.columns!r}\n{right.columns!r}"
            )
    
        if collect_input_frames:
            if check_dtype:  # check this _before_ we collect
                left_schema, right_schema = left.schema, right.schema
                if left_schema != right_schema:
                    raise_assert_detail(
                        objs, "lazy schemas are not equal", left_schema, right_schema
                    )
            left, right = left.collect(), right.collect()  # type: ignore[union-attr]
    
        if left.shape[0] != right.shape[0]:  # type: ignore[union-attr]
            raise_assert_detail(objs, "length mismatch", left.shape, right.shape)  # type: ignore[union-attr]
    
        if not check_row_order:
            try:
                left = left.sort(by=left.columns)
                right = right.sort(by=left.columns)
            except ComputeError as exc:
                raise InvalidAssert(
                    "cannot set `check_row_order=False` on frame with unsortable columns"
                ) from exc
    
        # note: does not assume a particular column order
        for c in left.columns:
            try:
                _assert_series_inner(
                    left[c],  # type: ignore[arg-type, index]
                    right[c],  # type: ignore[arg-type, index]
                    check_dtype=check_dtype,
                    check_exact=check_exact,
                    atol=atol,
                    rtol=rtol,
                    nans_compare_equal=nans_compare_equal,
                    categorical_as_str=categorical_as_str,
                )
            except AssertionError as exc:
                msg = f"values for column {c!r} are different."
>               raise AssertionError(msg) from exc
E               AssertionError: values for column 'a' are different.

polars/testing/asserts.py:136: AssertionError
====================================================================== short test summary info =======================================================================
FAILED test_asserts.py::test_series - AssertionError: Series are different (value mismatch)
FAILED test_asserts.py::test_df - AssertionError: values for column 'a' are different.
========================================================================= 2 failed in 0.14s ==========================================================================
```

</details>

That's 554 lines of output for two very simple test failures. It's basically impossible to find where the error message for the first test begins when scrolling up. We should look into disabling the tracebacks within the assertion utils.

It's already MUCH better when running `pytest --tb=short`. So we should probably enable that for our own test suite.

<details>

```
======================================================================== test session starts =========================================================================
platform linux -- Python 3.11.2, pytest-7.4.0, pluggy-1.3.0
rootdir: /home/stijn/code/polars/py-polars
configfile: pyproject.toml
plugins: cov-4.1.0, hypothesis-6.87.1, xdist-3.3.1
collected 2 items                                                                                                                                                    

test_asserts.py FF                                                                                                                                             [100%]

============================================================================== FAILURES ==============================================================================
____________________________________________________________________________ test_series _____________________________________________________________________________
test_asserts.py:8: in test_series
    assert_series_equal(s1, s2)
polars/testing/asserts.py:268: in assert_series_equal
    _assert_series_inner(
polars/testing/asserts.py:444: in _assert_series_inner
    raise_assert_detail(
E   AssertionError: Series are different (value mismatch)
E   [left]:  [3]
E   [right]: [4]
______________________________________________________________________________ test_df _______________________________________________________________________________
polars/testing/asserts.py:124: in assert_frame_equal
    _assert_series_inner(
polars/testing/asserts.py:444: in _assert_series_inner
    raise_assert_detail(
E   AssertionError: Series are different (value mismatch)
E   [left]:  [3]
E   [right]: [4]

The above exception was the direct cause of the following exception:
test_asserts.py:14: in test_df
    assert_frame_equal(df1, df2)
polars/testing/asserts.py:136: in assert_frame_equal
    raise AssertionError(msg) from exc
E   AssertionError: values for column 'a' are different.
====================================================================== short test summary info =======================================================================
FAILED test_asserts.py::test_series - AssertionError: Series are different (value mismatch)
FAILED test_asserts.py::test_df - AssertionError: values for column 'a' are different.
========================================================================= 2 failed in 0.12s ==========================================================================
```

</details>

### Assertion details for value mismatch

We have nice assertion details, but they can be slightly misleading. Consider the value mismatch - it shows only a single value (3 vs 4). Yet each Series has three elements. It would be nicer to view both side-by-side with all elements (we can specify some maximum).

--- a/py-polars/polars/testing/asserts/frame.py
+++ b/py-polars/polars/testing/asserts/frame.py
@@ -127,10 +127,11 @@ def assert_frame_equal(
         left, right = _sort_dataframes(left, right)
 
     for c in left.columns:
+        s_left, s_right = left.get_column(c), right.get_column(c)
         try:
             _assert_series_values_equal(
-                left.get_column(c),
-                right.get_column(c),
+                s_left,
+                s_right,
                 check_exact=check_exact,
                 rtol=rtol,
                 atol=atol,
@@ -138,8 +139,13 @@ def assert_frame_equal(
                 categorical_as_str=categorical_as_str,
             )
         except AssertionError as exc:
-            msg = f"values for column {c!r} are different"
-            raise AssertionError(msg) from exc
+            raise_assertion_error(
+                objects,
+                f"value mismatch for column {c!r}",
+                s_left.to_list(),
+                s_right.to_list(),
+                cause=exc,
+            )
 
 
 def _assert_correct_input_type(

--- a/py-polars/polars/testing/asserts/series.py
+++ b/py-polars/polars/testing/asserts/series.py
@@ -170,15 +170,25 @@ def _assert_series_values_equal(
 
     # Check nested dtypes in separate function
     if _comparing_nested_numerics(left.dtype, right.dtype):
-        if _assert_series_nested(
-            left=left.filter(unequal),
-            right=right.filter(unequal),
-            check_exact=check_exact,
-            rtol=rtol,
-            atol=atol,
-            nans_compare_equal=nans_compare_equal,
-            categorical_as_str=categorical_as_str,
-        ):
+        try:
+            _assert_series_nested_values_equal(
+                left=left.filter(unequal),
+                right=right.filter(unequal),
+                check_exact=check_exact,
+                rtol=rtol,
+                atol=atol,
+                nans_compare_equal=nans_compare_equal,
+                categorical_as_str=categorical_as_str,
+            )
+        except AssertionError as exc:
+            raise_assertion_error(
+                "Series",
+                "nested value mismatch",
+                left=left.to_list(),
+                right=right.to_list(),
+                cause=exc,
+            )
+        else:  # All nested values match
             return
 
     # If no differences found during exact checking, we're done
@@ -192,10 +202,7 @@ def _assert_series_values_equal(
         or right.dtype not in NUMERIC_DTYPES
     ):
         raise_assertion_error(
-            "Series",
-            "exact value mismatch",
-            left=left.to_list(),
-            right=right.to_list(),
+            "Series", "exact value mismatch", left=left.to_list(), right=right.to_list()
         )
 
     _assert_series_null_values_match(left, right)
@@ -207,6 +214,47 @@ def _assert_series_values_equal(
         rtol=rtol,
         atol=atol,
     )
+
+
+def _assert_series_nested_values_equal(
+    left: Series,
+    right: Series,
+    *,
+    check_exact: bool,
+    rtol: float,
+    atol: float,
+    nans_compare_equal: bool,
+    categorical_as_str: bool,
+) -> None:
+    # compare nested lists element-wise
+    if _comparing_lists(left.dtype, right.dtype):
+        for s1, s2 in zip(left, right):
+            if s1 is None or s2 is None:
+                raise_assertion_error("Series", "nested value mismatch", s1, s2)
+
+            _assert_series_values_equal(
+                s1,
+                s2,
+                check_exact=check_exact,
+                rtol=rtol,
+                atol=atol,
+                nans_compare_equal=nans_compare_equal,
+                categorical_as_str=categorical_as_str,
+            )
+
+    # unnest structs as series and compare
+    else:
+        ls, rs = left.struct.unnest(), right.struct.unnest()
+        for s1, s2 in zip(ls, rs):
+            _assert_series_values_equal(
+                s1,
+                s2,
+                check_exact=check_exact,
+                rtol=rtol,
+                atol=atol,
+                nans_compare_equal=nans_compare_equal,
+                categorical_as_str=categorical_as_str,
+            )
 
 
 def _assert_series_null_values_match(left: Series, right: Series) -> None:
@@ -240,53 +288,6 @@ def _assert_series_nan_values_match(
             left.to_list(),
             right.to_list(),
         )
-
-
-def _assert_series_nested(
-    left: Series,
-    right: Series,
-    *,
-    check_exact: bool,
-    rtol: float,
-    atol: float,
-    nans_compare_equal: bool,
-    categorical_as_str: bool,
-) -> bool:
-    # compare nested lists element-wise
-    if _comparing_lists(left.dtype, right.dtype):
-        for s1, s2 in zip(left, right):
-            if (s1 is None and s2 is not None) or (s2 is None and s1 is not None):
-                raise_assertion_error("Series", "nested value mismatch", s1, s2)
-
-            _assert_series_values_equal(
-                s1,
-                s2,
-                check_exact=check_exact,
-                rtol=rtol,
-                atol=atol,
-                nans_compare_equal=nans_compare_equal,
-                categorical_as_str=categorical_as_str,
-            )
-        return True
-
-    # unnest structs as series and compare
-    elif _comparing_structs(left.dtype, right.dtype):
-        ls, rs = left.struct.unnest(), right.struct.unnest()
-        for s1, s2 in zip(ls, rs):
-            _assert_series_values_equal(
-                s1,
-                s2,
-                check_exact=check_exact,
-                rtol=rtol,
-                atol=atol,
-                nans_compare_equal=nans_compare_equal,
-                categorical_as_str=categorical_as_str,
-            )
-        return True
-    else:
-        # fall-back to outer codepath (if mismatched dtypes we would expect
-        # the equality check to fail - unless ALL series values are null)
-        return False
 
 
 def _comparing_floats(left: PolarsDataType, right: PolarsDataType) -> bool:

--- a/py-polars/polars/testing/asserts/series.py
+++ b/py-polars/polars/testing/asserts/series.py
@@ -8,6 +8,8 @@ from polars.datatypes import (
     UNSIGNED_INTEGER_DTYPES,
     Array,
     Categorical,
+    Decimal,
+    Float64,
     Int64,
     List,
     Struct,
@@ -150,6 +152,14 @@ def _assert_series_values_equal(
             left = left.cast(Utf8)
         if right.dtype == Categorical:
             right = right.cast(Utf8)
+
+    # Handle decimals
+    # TODO: Delete this branch when Decimal equality is implemented
+    # https://github.com/pola-rs/polars/issues/12118
+    if left.dtype == Decimal:
+        left = left.cast(Float64)
+    if right.dtype == Decimal:
+        right = right.cast(Float64)
 
     # Determine unequal elements
     try:

--- a/py-polars/tests/unit/testing/test_assert_frame_equal.py
+++ b/py-polars/tests/unit/testing/test_assert_frame_equal.py
@@ -198,7 +198,7 @@ def test_compare_frame_equal_nans() -> None:
         schema=[("x", pl.Float32), ("y", pl.Float64)],
     )
     assert_frame_not_equal(df1, df2)
-    with pytest.raises(AssertionError, match="values for column 'y' are different"):
+    with pytest.raises(AssertionError, match="value mismatch for column 'y'"):
         assert_frame_equal(df1, df2, check_exact=True)
 
 
@@ -215,7 +215,7 @@ def test_compare_frame_equal_nested_nans() -> None:
         schema=[("x", pl.List(pl.Float32)), ("y", pl.List(pl.Float64))],
     )
     assert_frame_not_equal(df1, df2)
-    with pytest.raises(AssertionError, match="values for column 'y' are different"):
+    with pytest.raises(AssertionError, match="value mismatch for column 'y'"):
         assert_frame_equal(df1, df2, check_exact=True)
 
     # struct dtype
@@ -328,7 +328,7 @@ def test_assert_frame_equal_ignore_row_order() -> None:
     df1 = pl.DataFrame({"a": [1, 2], "b": [4, 3]})
     df2 = pl.DataFrame({"a": [2, 1], "b": [3, 4]})
     df3 = pl.DataFrame({"b": [3, 4], "a": [2, 1]})
-    with pytest.raises(AssertionError, match="values for column 'a' are different"):
+    with pytest.raises(AssertionError, match="value mismatch for column 'a'"):
         assert_frame_equal(df1, df2)
 
     assert_frame_equal(df1, df2, check_row_order=False)

--- a/py-polars/tests/unit/testing/test_assert_series_equal.py
+++ b/py-polars/tests/unit/testing/test_assert_series_equal.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 import math
 from datetime import datetime, time, timedelta
+from decimal import Decimal as D
 from typing import Any
 
 import pytest
@@ -620,6 +621,34 @@ def test_assert_series_equal_nested_int() -> None:
         assert_series_equal(s1, s2, check_exact=True)
 
 
+def test_series_equal_nested_lengths_mismatch() -> None:
+    s1 = pl.Series([[1.0, 2.0], [3.0, 4.0]], dtype=pl.List(pl.Float64))
+    s2 = pl.Series([[1.0, 2.0, 3.0], [4.0]], dtype=pl.List(pl.Float64))
+
+    with pytest.raises(AssertionError, match="nested value mismatch"):
+        assert_series_equal(s1, s2)
+
+
+def test_series_equal_decimals_exact() -> None:
+    s1 = pl.Series([D("1.00000"), D("2.00000")], dtype=pl.Decimal)
+    s2 = pl.Series([D("1.00000"), D("2.00001")], dtype=pl.Decimal)
+    with pytest.raises(AssertionError, match="exact value mismatch"):
+        assert_series_equal(s1, s2, check_exact=True)
+
+
+def test_series_equal_decimals_inexact() -> None:
+    s1 = pl.Series([D("1.00000"), D("2.00000")], dtype=pl.Decimal)
+    s2 = pl.Series([D("1.00000"), D("2.00001")], dtype=pl.Decimal)
+    assert_series_equal(s1, s2, check_exact=False)
+
+
+def test_series_equal_decimals_inexact_fail() -> None:
+    s1 = pl.Series([D("1.00000"), D("2.00000")], dtype=pl.Decimal)
+    s2 = pl.Series([D("1.00000"), D("2.00001")], dtype=pl.Decimal)
+    with pytest.raises(AssertionError, match="value mismatch"):
+        assert_series_equal(s1, s2, check_exact=False, rtol=0)
+
+
 def test_compare_series_nans_assert_equal_deprecated() -> None:
     srs1 = pl.Series([1.0, 2.0, nan, 4.0, None, 6.0])
     srs2 = pl.Series([1.0, nan, 3.0, 4.0, None, 6.0])
@@ -765,11 +794,3 @@ def test_assert_series_equal_raises_assertion_error_deprecated_nans_compare_equa
         assert_series_equal(s1, s2, nans_compare_equal=False, **kwargs)
     with pytest.deprecated_call():
         assert_series_not_equal(s1, s2, nans_compare_equal=False, **kwargs)
-
-
-def test_series_equal_nested_lengths_mismatch():
-    s1 = pl.Series([[1.0, 2.0], [3.0, 4.0]], dtype=pl.List(pl.Float64))
-    s2 = pl.Series([[1.0, 2.0, 3.0], [4.0]], dtype=pl.List(pl.Float64))
-
-    with pytest.raises(AssertionError, match="nested value mismatch"):
-        assert_series_equal(s1, s2)

--- a/py-polars/tests/unit/testing/test_assert_series_equal.py
+++ b/py-polars/tests/unit/testing/test_assert_series_equal.py
@@ -561,9 +561,26 @@ def test_assert_series_equal_nested_struct_float() -> None:
         assert_series_equal(s1, s2)
 
 
-def test_assert_series_equal_nested_list_full_null() -> None:
+def test_assert_series_equal_full_null_incompatible_dtypes_raises() -> None:
+    s1 = pl.Series([None, None], dtype=pl.Categorical)
+    s2 = pl.Series([None, None], dtype=pl.Int16)
+
+    # You could argue this should pass, but it's rare enough not to warrant the
+    # additional check
+    with pytest.raises(AssertionError, match="incompatible data types"):
+        assert_series_equal(s1, s2, check_dtype=False)
+
+
+def test_assert_series_equal_full_null_nested_list() -> None:
     s = pl.Series([None, None], dtype=pl.List(pl.Float64))
     assert_series_equal(s, s)
+
+
+def test_assert_series_equal_full_null_nested_not_nested() -> None:
+    s1 = pl.Series([None, None], dtype=pl.List(pl.Float64))
+    s2 = pl.Series([None, None], dtype=pl.Float64)
+
+    assert_series_equal(s1, s2, check_dtype=False)
 
 
 def test_assert_series_equal_nested_list_nan() -> None:
@@ -576,13 +593,6 @@ def test_assert_series_equal_nested_list_none() -> None:
     s2 = pl.Series([[1.0, 2.0], None], dtype=pl.List(pl.Float64))
 
     assert_series_equal(s1, s2)
-
-
-def test_assert_series_equal_full_none_nested_not_nested() -> None:
-    s1 = pl.Series([None, None], dtype=pl.List(pl.Float64))
-    s2 = pl.Series([None, None], dtype=pl.Float64)
-
-    assert_series_equal(s1, s2, check_dtype=False)
 
 
 def test_assert_series_equal_unsigned_ints_underflow() -> None:
@@ -755,3 +765,11 @@ def test_assert_series_equal_raises_assertion_error_deprecated_nans_compare_equa
         assert_series_equal(s1, s2, nans_compare_equal=False, **kwargs)
     with pytest.deprecated_call():
         assert_series_not_equal(s1, s2, nans_compare_equal=False, **kwargs)
+
+
+def test_series_equal_nested_lengths_mismatch():
+    s1 = pl.Series([[1.0, 2.0], [3.0, 4.0]], dtype=pl.List(pl.Float64))
+    s2 = pl.Series([[1.0, 2.0, 3.0], [4.0]], dtype=pl.List(pl.Float64))
+
+    with pytest.raises(AssertionError, match="nested value mismatch"):
+        assert_series_equal(s1, s2)

